MantidQt/MantidWidgets/src/SlitCalculator.cpp

```diff
@@ -1,48 +1,122 @@
 #include "MantidQtMantidWidgets/SlitCalculator.h"
 #include "MantidAPI/AlgorithmManager.h"
-#include <math.h>
+#include "MantidAPI/Progress.h"
+#include "MantidGeometry/Instrument/InstrumentDefinitionParser.h"
 
-namespace MantidQt
-{
-  namespace MantidWidgets
-  {
-    SlitCalculator::SlitCalculator(QWidget *parent)
-    {
-      Q_UNUSED(parent);
-      ui.setupUi(this);
-      on_recalculate_triggered();
-    }
+#include <math.h>
 
-    SlitCalculator::~SlitCalculator()
-    {
-    }
+namespace MantidQt {
+namespace MantidWidgets {
+SlitCalculator::SlitCalculator(QWidget *parent) {
+  Q_UNUSED(parent);
+  ui.setupUi(this);
+  if (currentInstrumentName == "") {
+    // set up the initial instrument if there is not one associated
+    // with slit calculator
+    currentInstrumentName = "INTER";
+    setInstrument(currentInstrumentName);
+  }
+  on_recalculate_triggered();
+}
+void SlitCalculator::processInstrumentHasBeenChanged() {
+  // used in refl main window to indicate that slitCalculator fields
+  // need to update because another instrument has been selected.
+  on_recalculate_triggered();
+}
+SlitCalculator::~SlitCalculator() {}
+void SlitCalculator::setInstrument(std::string instrumentName) {
+  // we want to get the most up-to-date definition, so we use the current
+  // date/time
+  auto date = Mantid::Kernel::DateAndTime::getCurrentTime().toISO8601String();
+  // find the full path to the definition file
+  auto filename =
+      Mantid::API::ExperimentInfo::getInstrumentFilename(instrumentName, date);
+  // parse the XML that we have found for the definition
+  Mantid::Geometry::InstrumentDefinitionParser parser =
+      Mantid::Geometry::InstrumentDefinitionParser(
+          filename, instrumentName,
+          Mantid::Kernel::Strings::loadFile(filename));
+  // retrieving the mangled name of the instrument
+  std::string instrumentNameMangled = parser.getMangledName();
+  // See if we have a definition already in the InstrumentDataService
+  if (Mantid::API::InstrumentDataService::Instance().doesExist(
+          instrumentNameMangled)) {
+    // If it does, set the associated instrument to the one we have found.
+    this->instrument = Mantid::API::InstrumentDataService::Instance().retrieve(
+        instrumentNameMangled);
+  } else {
+    // We set the instrument from XML that we have found.
+    Mantid::API::Progress *prog = new Mantid::API::Progress();
```

Comment:
Would have been best to keep Progress on the stack to avoid the delete. That'd have kept this exception safe.